feat: Smart Pointers

TAdev0 opened the pull request:
Closes #236 

@enitrat  I think I need a bit of guidance here, this is why i push this PR even though not entirely finished.

In the issue , you commented:
 "... dive deeper by compiling Cairo0 code using pointers to CASM, then do the equivalent code in Cairo 1 with Boxes/Nullable to understand how the memory is managed"
 
 I checked a few short example, looking at Sierra and Casm.  I struggle to understand and see in the execution trace what really happens under the hood, like the allocation of a new memory segment, to show it clearly in an example.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cairo-book/cairo-book/750)
<!-- Reviewable:end -->

TAdev0 commented:
@enitrat  I also think that we should move this chapter just after inlining. Indeed, in inlining chapter we introduced Sierra files structure, and here i use a sierra file again. Seems more logic to have this chapter after

enitrat reviewed (changes requested):
remember: the book focuses on explaining _concepts_ and how to apply these concepts, rather than purely describing an API

a good rule of thumb is : What - Why - How
What is this ? Why would I use it ? How can I use it ?

> Reviewed 3 of 6 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cairo-book/cairo-book/750#-:-NwOmpn8EYPZouzwSloc:bmd121r)* status: 3 of 6 files reviewed, 14 unresolved discussions (waiting on @TAdev0)

___
*[`src/ch11-03-smart-pointers.md` line 9 at r1](https://reviewable.io/reviews/cairo-book/cairo-book/750#-NwOieK34VJggxavc--S:-NwOieK34VJggxavc--T:bu8h1o5) ([raw file](https://github.com/cairo-book/cairo-book/blob/7647b54459413f133a6681fd746f7fce039d6124/src/ch11-03-smart-pointers.md#L9)):*
> ```Markdown
> In the specific case of Cairo, smart pointers play a crucial role in preventing illegal memory address references. Their usage ensures that dereferences cannot fail. Remember that it is not possible to prove the execution of a code that fails with the Cairo VM. Hence, it is mandatory to make sure that failing attempt to dereference a pointer and access an unallocated memory cell will never happen.
> 
> ## The `Box` Type
> ```

The title should give information about what the Box type does. If I fast-read the book I should be able to understand concepts from the title only
___
*[`src/ch11-03-smart-pointers.md` line 11 at r1](https://reviewable.io/reviews/cairo-book/cairo-book/750#-NwOivwI0qGE4s7RwYBy:-NwOivwI0qGE4s7RwYBz:bjcsek8) ([raw file](https://github.com/cairo-book/cairo-book/blob/7647b54459413f133a6681fd746f7fce039d6124/src/ch11-03-smart-pointers.md#L11)):*
> ```Markdown
> ## The `Box` Type
> 
> The `Box` type represents the main smart pointer type in Cairo. Inspired by Rust, it allows us to allocate a new memory segment for our type, and access this segment using a pointer that can only be manipulated in one place at a time.
> ```

I would remove this as it's quite different from Rust's box
___
*[`src/ch11-03-smart-pointers.md` line 11 at r1](https://reviewable.io/reviews/cairo-book/cairo-book/750#-NwOj5EW3ij7-jaAivFq:-NwOj5EW3ij7-jaAivFr:b7bb5ju) ([raw file](https://github.com/cairo-book/cairo-book/blob/7647b54459413f133a6681fd746f7fce039d6124/src/ch11-03-smart-pointers.md#L11)):*
> ```Markdown
> ## The `Box` Type
> 
> The `Box` type represents the main smart pointer type in Cairo. Inspired by Rust, it allows us to allocate a new memory segment for our type, and access this segment using a pointer that can only be manipulated in one place at a time.
> ```

 the type is only a high-level concept



_Suggestion:_
```Markdown
data
```
___
*[`src/ch11-03-smart-pointers.md` line 11 at r1](https://reviewable.io/reviews/cairo-book/cairo-book/750#-NwOkfdBBQDONQZziDog:-NwOkfdBBQDONQZziDoh:b-lcl4xv) ([raw file](https://github.com/cairo-book/cairo-book/blob/7647b54459413f133a6681fd746f7fce039d6124/src/ch11-03-smart-pointers.md#L11)):*
> ```Markdown
> ## The `Box` Type
> 
> The `Box` type represents the main smart pointer type in Cairo. Inspired by Rust, it allows us to allocate a new memory segment for our type, and access this segment using a pointer that can only be manipulated in one place at a time.
> ```

There is a single segment `boxed_segment` for all boxed values. You don't create a new segment each time; you just append to the boxed segment



_Code quote:_
```Markdown
it allows us to allocate a new memory segment for our type
```
___
*[`src/ch11-03-smart-pointers.md` line 13 at r1](https://reviewable.io/reviews/cairo-book/cairo-book/750#-NwOjQhu746ffeLFNgr2:-NwOjQhu746ffeLFNgr3:b45h) ([raw file](https://github.com/cairo-book/cairo-book/blob/7647b54459413f133a6681fd746f7fce039d6124/src/ch11-03-smart-pointers.md#L13)):*
> ```Markdown
> The `Box` type represents the main smart pointer type in Cairo. Inspired by Rust, it allows us to allocate a new memory segment for our type, and access this segment using a pointer that can only be manipulated in one place at a time.
> 
> The Cairo VM memory is composed avec segments defined by an index, each segment containing slots to store felts. Each time you instantiate a variable of `Box<T>` type, you allocate and initialize a pointer instance in a new memory segment for this pointer variable.
> ```





_Suggestion:_
```Markdown
of
```
___
*[`src/ch11-03-smart-pointers.md` line 15 at r1](https://reviewable.io/reviews/cairo-book/cairo-book/750#-NwOjYCy-iY8HKlLnnsZ:-NwOjYCy-iY8HKlLnns_:b-busffr) ([raw file](https://github.com/cairo-book/cairo-book/blob/7647b54459413f133a6681fd746f7fce039d6124/src/ch11-03-smart-pointers.md#L15)):*
> ```Markdown
> The Cairo VM memory is composed avec segments defined by an index, each segment containing slots to store felts. Each time you instantiate a variable of `Box<T>` type, you allocate and initialize a pointer instance in a new memory segment for this pointer variable.
> 
> ### `BoxTrait` and its methods
> ```

I would skip this, the goal is not to be a reference of the core library.
___
*[`src/ch11-03-smart-pointers.md` line 43 at r1](https://reviewable.io/reviews/cairo-book/cairo-book/750#-NwOl4ud9cHg9nzhKQVy:-NwOl4ud9cHg9nzhKQVz:bld14o0) ([raw file](https://github.com/cairo-book/cairo-book/blob/7647b54459413f133a6681fd746f7fce039d6124/src/ch11-03-smart-pointers.md#L43)):*
> ```Markdown
> The important thing to note here is that in the case of `const_passed_by_value`, the compiler actually creates a copy of the  `RandomStruct` with all its fields when passing the variable to the function, while `box_const_passed_by_value` only requires the copy of the value of the pointer which is a felt.
> 
> #### Sierra Code
> ```

I don't think Sierra code is interesting here; what matters is the actual instructions executed by the Cairo VM.  
___
*[`src/ch11-03-smart-pointers.md` line 58 at r1](https://reviewable.io/reviews/cairo-book/cairo-book/750#-NwOlJ51CjDznjrX3ON8:-NwOlJ51CjDznjrX3ON9:b1aqhva) ([raw file](https://github.com/cairo-book/cairo-book/blob/7647b54459413f133a6681fd746f7fce039d6124/src/ch11-03-smart-pointers.md#L58)):*
> ```Markdown
> - Drop the previouslys unboxed variable of type `RandomStruct`.
> 
> ### Quick Recap
> ```

This should be in the introduction of the box type: before learning how to use them I should why I would use them 
___
*[`src/ch11-03-smart-pointers.md` line 60 at r1](https://reviewable.io/reviews/cairo-book/cairo-book/750#-NwOlYPeETjHLbX98E72:-NwOlYPeETjHLbX98E73:undefined) ([raw file](https://github.com/cairo-book/cairo-book/blob/7647b54459413f133a6681fd746f7fce039d6124/src/ch11-03-smart-pointers.md#L60)):*
> ```Markdown
> ### Quick Recap
> 
> Using boxes can improve efficiency as it allows to pass the variable, which is a reference, from a function to another without the need to copy the whole value. Indeed, there is no access to the value contained in the box while there is no unbox. We still need to pay to copy the value of the reference itself, but not the whole `const` struct. Because there is a dedicated segment and because a box can only be used in one place at a time, the `Box<T>` type reduces the number of steps while ensuring soundness in pointers dereferences.
> ```


___
*[`src/ch11-03-smart-pointers.md` line 60 at r1](https://reviewable.io/reviews/cairo-book/cairo-book/750#-NwOle3RAcjVWNafC7m4:-NwOle3RAcjVWNafC7m5:b-di1uei) ([raw file](https://github.com/cairo-book/cairo-book/blob/7647b54459413f133a6681fd746f7fce039d6124/src/ch11-03-smart-pointers.md#L60)):*
> ```Markdown
> ### Quick Recap
> 
> Using boxes can improve efficiency as it allows to pass the variable, which is a reference, from a function to another without the need to copy the whole value. Indeed, there is no access to the value contained in the box while there is no unbox. We still need to pay to copy the value of the reference itself, but not the whole `const` struct. Because there is a dedicated segment and because a box can only be used in one place at a time, the `Box<T>` type reduces the number of steps while ensuring soundness in pointers dereferences.
> ```

be careful with the word reference here. I would avoid mixing terminology



_Code quote:_
```Markdown
which is a reference
```
___
*[`src/ch11-03-smart-pointers.md` line 60 at r1](https://reviewable.io/reviews/cairo-book/cairo-book/750#-NwOls0cB2al4ZQCt-9Y:-NwOls0cB2al4ZQCt-9Z:b-d16fq1) ([raw file](https://github.com/cairo-book/cairo-book/blob/7647b54459413f133a6681fd746f7fce039d6124/src/ch11-03-smart-pointers.md#L60)):*
> ```Markdown
> ### Quick Recap
> 
> Using boxes can improve efficiency as it allows to pass the variable, which is a reference, from a function to another without the need to copy the whole value. Indeed, there is no access to the value contained in the box while there is no unbox. We still need to pay to copy the value of the reference itself, but not the whole `const` struct. Because there is a dedicated segment and because a box can only be used in one place at a time, the `Box<T>` type reduces the number of steps while ensuring soundness in pointers dereferences.
> ```

I don't think it conveys the idea correctly, at least after reading this im quite confused.

What you want to outline is that you would use boxes when you want to pass large amounts of data to a function, and want to ensure that the data won't be copied when doing so for efficiency purposes.



_Code quote:_
```Markdown
Using boxes can improve efficiency as it allows to pass the variable, which is a reference, from a function to another without the need to copy the whole value. Indeed, there is no access to the value contained in the box while there is no unbox. We still need to pay to copy the value of the reference itself, but not the whole `const` struct. Because there is a dedicated segment and because a box can only be used in one place at a time, the `Box<T>` type reduces the number of steps while ensuring soundness in pointers dereferences.
```
___
*[`src/ch11-03-smart-pointers.md` line 63 at r1](https://reviewable.io/reviews/cairo-book/cairo-book/750#-NwOmWYW4T1cR2GGz9BR:-NwOmWYW4T1cR2GGz9BS:b3zzgmy) ([raw file](https://github.com/cairo-book/cairo-book/blob/7647b54459413f133a6681fd746f7fce039d6124/src/ch11-03-smart-pointers.md#L63)):*
> ```Markdown
> 
> Also, the `Box<T>` is of paramount importance for memory safety, especially when dealing with arrays. Because arrays can grow in size with the `append` method, there is a room for potential illegal out of bounds access that would induce a failure. This is why a new memory segment is allocated when a new array is instantiated. Hence, if you want to retrieve the value of an item contained in an array, you will get at some point a `Box<@T>` that you will have to unbox.
> ```

Also: when you have a type whose size cannot be known at compile time, like recursive structs
___
*[`src/ch11-03-smart-pointers.md` line 64 at r1](https://reviewable.io/reviews/cairo-book/cairo-book/750#-NwOmajh5ZlR63fQBFui:-NwOmajh5ZlR63fQBFuj:b-a99xwj) ([raw file](https://github.com/cairo-book/cairo-book/blob/7647b54459413f133a6681fd746f7fce039d6124/src/ch11-03-smart-pointers.md#L64)):*
> ```Markdown
> Also, the `Box<T>` is of paramount importance for memory safety, especially when dealing with arrays. Because arrays can grow in size with the `append` method, there is a room for potential illegal out of bounds access that would induce a failure. This is why a new memory segment is allocated when a new array is instantiated. Hence, if you want to retrieve the value of an item contained in an array, you will get at some point a `Box<@T>` that you will have to unbox.
> 
> ## The `Nullable<T>` Type
> ```

The title should indicate the "value proposal" of the Nullable type
___
*[`src/ch11-03-smart-pointers.md` line 68 at r1](https://reviewable.io/reviews/cairo-book/cairo-book/750#-NwOmZSX-nkSDcyz4R9O:-NwOmZSX-nkSDcyz4R9P:btbrdae) ([raw file](https://github.com/cairo-book/cairo-book/blob/7647b54459413f133a6681fd746f7fce039d6124/src/ch11-03-smart-pointers.md#L68)):*
> ```Markdown
> `Nullable<T>` is another type of smart pointer that can either point to a value or be null in the absence of value. It is usually used in Object Oriented Programming Languages when a reference doesn't point anywhere. The difference with `Option<T>` is that the wrapped value is stored inside a `Box<T>` smart pointer data type.
> 
> ### Usage
> ```

not interesting


<!-- Sent from Reviewable.io -->

TAdev0 reviewed (commented):
yeah i know... but i'm so unused to unbox that it is hard for me to have a natural feeling in explaining correctly the concept and applying them. Will improve that. I feel like there are 2 main use cases : 
- usage in arrays, for safety reason, this is enforced
- optional usage, for efficiency

is this correct?

> Reviewed 1 of 6 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cairo-book/cairo-book/750#-:-NwOtpHnE3tldsigzckj:b-hvnqnc)* status: 4 of 6 files reviewed, 14 unresolved discussions (waiting on @enitrat)

___
*[`src/ch11-03-smart-pointers.md` line 11 at r1](https://reviewable.io/reviews/cairo-book/cairo-book/750#-NwOj5EW3ij7-jaAivFq:-NwOuz6z9ErRh3e-gc0A:b2ena1w) ([raw file](https://github.com/cairo-book/cairo-book/blob/7647b54459413f133a6681fd746f7fce039d6124/src/ch11-03-smart-pointers.md#L11)):*
<details><summary><i>Previously, enitrat (Mathieu) wrote…</i></summary><blockquote>

the type is only a high-level concept
</blockquote></details>

true, nice
___
*[`src/ch11-03-smart-pointers.md` line 11 at r1](https://reviewable.io/reviews/cairo-book/cairo-book/750#-NwOkfdBBQDONQZziDog:-NwOvD75BMvEgFFk6nZs:b-a7knop) ([raw file](https://github.com/cairo-book/cairo-book/blob/7647b54459413f133a6681fd746f7fce039d6124/src/ch11-03-smart-pointers.md#L11)):*
<details><summary><i>Previously, enitrat (Mathieu) wrote…</i></summary><blockquote>

There is a single segment `boxed_segment` for all boxed values. You don't create a new segment each time; you just append to the boxed segment
</blockquote></details>

didnt know about that. Thanks!!!
___
*[`src/ch11-03-smart-pointers.md` line 11 at r1](https://reviewable.io/reviews/cairo-book/cairo-book/750#-NwOivwI0qGE4s7RwYBy:-NwOwcdY8c3h96dtN8RN:b-896fix) ([raw file](https://github.com/cairo-book/cairo-book/blob/7647b54459413f133a6681fd746f7fce039d6124/src/ch11-03-smart-pointers.md#L11)):*
<details><summary><i>Previously, enitrat (Mathieu) wrote…</i></summary><blockquote>

I would remove this as it's quite different from Rust's box
</blockquote></details>

Done.
___
*[`src/ch11-03-smart-pointers.md` line 13 at r1](https://reviewable.io/reviews/cairo-book/cairo-book/750#-NwOjQhu746ffeLFNgr2:-NwOx-oiC7YXZNajmR9M:bhldvlx) ([raw file](https://github.com/cairo-book/cairo-book/blob/7647b54459413f133a6681fd746f7fce039d6124/src/ch11-03-smart-pointers.md#L13)):*
> ```Markdown
> The `Box` type represents the main smart pointer type in Cairo. Inspired by Rust, it allows us to allocate a new memory segment for our type, and access this segment using a pointer that can only be manipulated in one place at a time.
> 
> The Cairo VM memory is composed avec segments defined by an index, each segment containing slots to store felts. Each time you instantiate a variable of `Box<T>` type, you allocate and initialize a pointer instance in a new memory segment for this pointer variable.
> ```

lol thanks
___
*[`src/ch11-03-smart-pointers.md` line 58 at r1](https://reviewable.io/reviews/cairo-book/cairo-book/750#-NwOlJ51CjDznjrX3ON8:-NwP0Dxd2117I6Yz4N4P:b-896fix) ([raw file](https://github.com/cairo-book/cairo-book/blob/7647b54459413f133a6681fd746f7fce039d6124/src/ch11-03-smart-pointers.md#L58)):*
<details><summary><i>Previously, enitrat (Mathieu) wrote…</i></summary><blockquote>

This should be in the introduction of the box type: before learning how to use them I should why I would use them
</blockquote></details>

Done.
___
*[`src/ch11-03-smart-pointers.md` line 60 at r1](https://reviewable.io/reviews/cairo-book/cairo-book/750#-NwOle3RAcjVWNafC7m4:-NwP0FCI67M6mLU9Xdhz:b-896fix) ([raw file](https://github.com/cairo-book/cairo-book/blob/7647b54459413f133a6681fd746f7fce039d6124/src/ch11-03-smart-pointers.md#L60)):*
<details><summary><i>Previously, enitrat (Mathieu) wrote…</i></summary><blockquote>

be careful with the word reference here. I would avoid mixing terminology
</blockquote></details>

Done.
___
*[`src/ch11-03-smart-pointers.md` line 60 at r1](https://reviewable.io/reviews/cairo-book/cairo-book/750#-NwOlYPeETjHLbX98E72:-NwP0FeAAZvpvyAondhP:b-896fix) ([raw file](https://github.com/cairo-book/cairo-book/blob/7647b54459413f133a6681fd746f7fce039d6124/src/ch11-03-smart-pointers.md#L60)):*
> ```Markdown
> ### Quick Recap
> 
> Using boxes can improve efficiency as it allows to pass the variable, which is a reference, from a function to another without the need to copy the whole value. Indeed, there is no access to the value contained in the box while there is no unbox. We still need to pay to copy the value of the reference itself, but not the whole `const` struct. Because there is a dedicated segment and because a box can only be used in one place at a time, the `Box<T>` type reduces the number of steps while ensuring soundness in pointers dereferences.
> ```

Done.
___
*[`src/ch11-03-smart-pointers.md` line 63 at r1](https://reviewable.io/reviews/cairo-book/cairo-book/750#-NwOmWYW4T1cR2GGz9BR:-NwP1wyYCWqXj4NiVroL:b-896fix) ([raw file](https://github.com/cairo-book/cairo-book/blob/7647b54459413f133a6681fd746f7fce039d6124/src/ch11-03-smart-pointers.md#L63)):*
<details><summary><i>Previously, enitrat (Mathieu) wrote…</i></summary><blockquote>

Also: when you have a type whose size cannot be known at compile time, like recursive structs
</blockquote></details>

Done.
___
*[`src/ch11-03-smart-pointers.md` line 64 at r1](https://reviewable.io/reviews/cairo-book/cairo-book/750#-NwOmajh5ZlR63fQBFui:-NwP4lduENmt5uYkZVf5:b-896fix) ([raw file](https://github.com/cairo-book/cairo-book/blob/7647b54459413f133a6681fd746f7fce039d6124/src/ch11-03-smart-pointers.md#L64)):*
<details><summary><i>Previously, enitrat (Mathieu) wrote…</i></summary><blockquote>

The title should indicate the "value proposal" of the Nullable type
</blockquote></details>

Done.
___
*[`src/ch11-03-smart-pointers.md` line 68 at r1](https://reviewable.io/reviews/cairo-book/cairo-book/750#-NwOmZSX-nkSDcyz4R9O:-NwP1jciA0pZe5tDynDT:b-cstnkg) ([raw file](https://github.com/cairo-book/cairo-book/blob/7647b54459413f133a6681fd746f7fce039d6124/src/ch11-03-smart-pointers.md#L68)):*
<details><summary><i>Previously, enitrat (Mathieu) wrote…</i></summary><blockquote>

not interesting
</blockquote></details>

I agree


<!-- Sent from Reviewable.io -->

TAdev0 reviewed (commented):
@enitrat  in dictionaries currently there is this : 

 The `Box<T>` type, inspired by Rust, allows us to allocate a new memory segment for our type, and access this segment using a pointer that can only be manipulated in one place at a time

i actually copied pasted this. So I guess this is wrong and i should modify that

> *[Reviewable](https://reviewable.io/reviews/cairo-book/cairo-book/750#-:-NwP6YKv2AFL6PZ5uhGq:b-p7kz91)* status: 4 of 6 files reviewed, 14 unresolved discussions (waiting on @enitrat)



<!-- Sent from Reviewable.io -->

enitrat reviewed (changes requested):
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cairo-book/cairo-book/750)* status: 3 of 7 files reviewed, 12 unresolved discussions (waiting on @TAdev0)

___
*[`src/ch03-02-dictionaries.md` line 184 at r2](https://reviewable.io/reviews/cairo-book/cairo-book/750#-NwTqqP9CgDIuj8WLqxO:-NwTqqP9CgDIuj8WLqxP:b45h) ([raw file](https://github.com/cairo-book/cairo-book/blob/fe1172b9b6e9a9570375296ee7d12add87cef05b/src/ch03-02-dictionaries.md#L184)):*
> ```Markdown
> To compensate this, you can wrap your type inside a `Nullable<T>`.
> 
> `Nullable<T>` is a smart pointer type that can either point to a value or be `null` in the absence of value. It is usually used in Object Oriented Programming Languages when a reference doesn't point anywhere. The difference with `Option` is that the wrapped value is stored inside a `Box<T>` data type. The `Box<T>` type is a smart pointer that allows us to use a dedicated `boxed_segment` memory segment for our type, and access this segment using a pointer that can only be manipulated in one place at a time. See [Smart Pointers Chapter](./ch11-03-smart-pointers.md) for more information.
> ```





_Suggestion:_
```Markdown
data
```
___
*[`src/ch11-03-smart-pointers.md` line 5 at r2](https://reviewable.io/reviews/cairo-book/cairo-book/750#-NwTr3kJEyyIYWVuJ-ts:-NwTr3kJEyyIYWVuJ-tt:bql2z1w) ([raw file](https://github.com/cairo-book/cairo-book/blob/fe1172b9b6e9a9570375296ee7d12add87cef05b/src/ch11-03-smart-pointers.md#L5)):*
> ```Markdown
> A pointer is a general concept for a variable that contains an address in memory. This address refers to, or “points at,” some other data.
> 
> Smart pointers, on the other hand, are data structures that act like a pointer but also have additional metadata and capabilities. The concept of smart pointers isn’t unique to Cairo: smart pointers originated in C++ and exist in other languages as well. For example, Rust has a variety of smart pointers defined in the standard library that provide various functionalities.
> ```

"a variety of smart pointers"
"provide various functionalities"

this doesn't give any meaningful information



_Code quote:_
```Markdown
For example, Rust has a variety of smart pointers defined in the standard library that provide various functionalities.
```
___
*[`src/ch11-03-smart-pointers.md` line 11 at r2](https://reviewable.io/reviews/cairo-book/cairo-book/750#-NwTrMjJ6t26t9pOaU6q:-NwTrMjJ6t26t9pOaU6r:b-qpylgv) ([raw file](https://github.com/cairo-book/cairo-book/blob/fe1172b9b6e9a9570375296ee7d12add87cef05b/src/ch11-03-smart-pointers.md#L11)):*
> ```Markdown
> The Cairo VM memory is composed with segments defined by an index, each segment containing slots to store felts. Using smart pointers allows one to store data in a dedicated segment of the memory, while manipulating a pointer variable in the regular segment.
> 
> ## The `Box<T>` Type to Safely Store Data
> ```

Very misleading imo. 



_Suggestion:_
```Markdown
The `Box<T>` Type to manipulate pointers
```
___
*[`src/ch11-03-smart-pointers.md` line 15 at r2](https://reviewable.io/reviews/cairo-book/cairo-book/750#-NwTrZHo22Ac8eBfd2gE:-NwTrZHo22Ac8eBfd2gF:undefined) ([raw file](https://github.com/cairo-book/cairo-book/blob/fe1172b9b6e9a9570375296ee7d12add87cef05b/src/ch11-03-smart-pointers.md#L15)):*
> ```Markdown
> The `Box` type represents the main smart pointer type in Cairo. It allows us to store data in the `boxed_segment` segment of the Cairo VM memory. This segment is dedicated to the storage of all boxed values. Whenever you instantiate a new pointer variable of type `Box<T>`, you append the data of type `T` to the boxed segment.
> 
> Using boxes can improve efficiency as it allows to pass the variable, which is a pointer, from a function to another without the need to copy the whole value. Indeed, there is no access to the value contained in the box while there is no unbox. We still need to execute steps to copy the value of the pointer itself, but not the value the box points to. If the variable stored in the box is very complex, improvement can be massive. Because there is a dedicated segment for boxes and because a box can only be used in one place at a time, using the `Box<T>` type when manipulating complex data types reduces the number of steps while ensuring soundness in pointers dereferences.
> ```





_Suggestion:_
```Markdown
a pointer to some data from a function to another without the need to copy the entire data..
```
___
*[`src/ch11-03-smart-pointers.md` line 15 at r2](https://reviewable.io/reviews/cairo-book/cairo-book/750#-NwTrld2Ct_3aAKqO7Bl:-NwTrld2Ct_3aAKqO7Bm:blw68ro) ([raw file](https://github.com/cairo-book/cairo-book/blob/fe1172b9b6e9a9570375296ee7d12add87cef05b/src/ch11-03-smart-pointers.md#L15)):*
> ```Markdown
> The `Box` type represents the main smart pointer type in Cairo. It allows us to store data in the `boxed_segment` segment of the Cairo VM memory. This segment is dedicated to the storage of all boxed values. Whenever you instantiate a new pointer variable of type `Box<T>`, you append the data of type `T` to the boxed segment.
> 
> Using boxes can improve efficiency as it allows to pass the variable, which is a pointer, from a function to another without the need to copy the whole value. Indeed, there is no access to the value contained in the box while there is no unbox. We still need to execute steps to copy the value of the pointer itself, but not the value the box points to. If the variable stored in the box is very complex, improvement can be massive. Because there is a dedicated segment for boxes and because a box can only be used in one place at a time, using the `Box<T>` type when manipulating complex data types reduces the number of steps while ensuring soundness in pointers dereferences.
> ```

Instead of having to write `n' values into memory before you call a function, you'd only have to write a single value that corresponds to your data pointer.



_Code quote:_
```Markdown
We still need to execute steps to copy the value of the pointer itself, but not the value the box points to. If the variable stored in the box is very complex, improvement can be massive.
```
___
*[`src/ch11-03-smart-pointers.md` line 15 at r2](https://reviewable.io/reviews/cairo-book/cairo-book/750#-NwTs8OR2BORLvWLsTuP:-NwTs8OR2BORLvWLsTuQ:b-4nr4gd) ([raw file](https://github.com/cairo-book/cairo-book/blob/fe1172b9b6e9a9570375296ee7d12add87cef05b/src/ch11-03-smart-pointers.md#L15)):*
> ```Markdown
> The `Box` type represents the main smart pointer type in Cairo. It allows us to store data in the `boxed_segment` segment of the Cairo VM memory. This segment is dedicated to the storage of all boxed values. Whenever you instantiate a new pointer variable of type `Box<T>`, you append the data of type `T` to the boxed segment.
> 
> Using boxes can improve efficiency as it allows to pass the variable, which is a pointer, from a function to another without the need to copy the whole value. Indeed, there is no access to the value contained in the box while there is no unbox. We still need to execute steps to copy the value of the pointer itself, but not the value the box points to. If the variable stored in the box is very complex, improvement can be massive. Because there is a dedicated segment for boxes and because a box can only be used in one place at a time, using the `Box<T>` type when manipulating complex data types reduces the number of steps while ensuring soundness in pointers dereferences.
> ```

Confusing: you're confusing the safety (of the Box type) and the efficiency properties (of pointers)



_Code quote:_
```Markdown
Because there is a dedicated segment for boxes and because a box can only be used in one place at a time, using the `Box<T>` type when manipulating complex data types reduces the number of steps while ensuring soundness in pointers dereferences.
```
___
*[`src/ch11-03-smart-pointers.md` line 22 at r2](https://reviewable.io/reviews/cairo-book/cairo-book/750#-NwTsJFU1PnYKb5UEMHa:-NwTsJFU1PnYKb5UEMHb:b-rgx4i) ([raw file](https://github.com/cairo-book/cairo-book/blob/fe1172b9b6e9a9570375296ee7d12add87cef05b/src/ch11-03-smart-pointers.md#L22)):*
> ```Markdown
> > - Efficiency: Manually defining variables of type `Box<T>` that contain a large amount of data can be very useful in reducing the number of steps as it ensures that the data contained in the box won't be copied when passed from a function to another.
> > 
> > - Safety: variable whose size is not known at compile time need to use the `Box<T` type for items they contain. For example, arrays use by default the `Box<T>` for items they contain. 
> ```

You always want to start sections with less technical content that introduces you to the concept before diving in.



_Code quote:_
```Markdown
> To resume, the `Box<T>` type is mainly used for 2 purposes :
> - Efficiency: Manually defining variables of type `Box<T>` that contain a large amount of data can be very useful in reducing the number of steps as it ensures that the data contained in the box won't be copied when passed from a function to another.
>
> - Safety: variable whose size is not known at compile time need to use the `Box<T` type for items they contain. For example, arrays use by default the `Box<T>` for items they contain.
```
___
*[`src/ch11-03-smart-pointers.md` line 22 at r2](https://reviewable.io/reviews/cairo-book/cairo-book/750#-NwTsXSa7rURIttSm-aq:-NwTsXSa7rURIttSm-ar:bipcax3) ([raw file](https://github.com/cairo-book/cairo-book/blob/fe1172b9b6e9a9570375296ee7d12add87cef05b/src/ch11-03-smart-pointers.md#L22)):*
> ```Markdown
> > - Efficiency: Manually defining variables of type `Box<T>` that contain a large amount of data can be very useful in reducing the number of steps as it ensures that the data contained in the box won't be copied when passed from a function to another.
> > 
> > - Safety: variable whose size is not known at compile time need to use the `Box<T` type for items they contain. For example, arrays use by default the `Box<T>` for items they contain. 
> ```

do they?



_Code quote:_
```Markdown
arrays use by default the `Box<T>` for items they contain.
```


<!-- Sent from Reviewable.io -->

TAdev0 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cairo-book/cairo-book/750)* status: 3 of 11 files reviewed, 12 unresolved discussions (waiting on @enitrat)

___
*[`src/ch03-02-dictionaries.md` line 184 at r2](https://reviewable.io/reviews/cairo-book/cairo-book/750#-NwTqqP9CgDIuj8WLqxO:-NwTxWB85y30LNhdiGCS:b-l6wj6h) ([raw file](https://github.com/cairo-book/cairo-book/blob/fe1172b9b6e9a9570375296ee7d12add87cef05b/src/ch03-02-dictionaries.md#L184)):*
> ```Markdown
> To compensate this, you can wrap your type inside a `Nullable<T>`.
> 
> `Nullable<T>` is a smart pointer type that can either point to a value or be `null` in the absence of value. It is usually used in Object Oriented Programming Languages when a reference doesn't point anywhere. The difference with `Option` is that the wrapped value is stored inside a `Box<T>` data type. The `Box<T>` type is a smart pointer that allows us to use a dedicated `boxed_segment` memory segment for our type, and access this segment using a pointer that can only be manipulated in one place at a time. See [Smart Pointers Chapter](./ch11-03-smart-pointers.md) for more information.
> ```

yes thanks
___
*[`src/ch11-03-smart-pointers.md` line 15 at r1](https://reviewable.io/reviews/cairo-book/cairo-book/750#-NwOjYCy-iY8HKlLnnsZ:-NwUDIjICTb9SppuwN4Y:b-896fix) ([raw file](https://github.com/cairo-book/cairo-book/blob/7647b54459413f133a6681fd746f7fce039d6124/src/ch11-03-smart-pointers.md#L15)):*
<details><summary><i>Previously, enitrat (Mathieu) wrote…</i></summary><blockquote>

I would skip this, the goal is not to be a reference of the core library.
</blockquote></details>

Done.
___
*[`src/ch11-03-smart-pointers.md` line 43 at r1](https://reviewable.io/reviews/cairo-book/cairo-book/750#-NwOl4ud9cHg9nzhKQVy:-NwUY5710vBeHpsnGdvR:b-lcgiwp) ([raw file](https://github.com/cairo-book/cairo-book/blob/7647b54459413f133a6681fd746f7fce039d6124/src/ch11-03-smart-pointers.md#L43)):*
<details><summary><i>Previously, enitrat (Mathieu) wrote…</i></summary><blockquote>

I don't think Sierra code is interesting here; what matters is the actual instructions executed by the Cairo VM.
</blockquote></details>

thats true
___
*[`src/ch11-03-smart-pointers.md` line 5 at r2](https://reviewable.io/reviews/cairo-book/cairo-book/750#-NwTr3kJEyyIYWVuJ-ts:-NwTy3128y0yxA2fLEve:bgkrc0j) ([raw file](https://github.com/cairo-book/cairo-book/blob/fe1172b9b6e9a9570375296ee7d12add87cef05b/src/ch11-03-smart-pointers.md#L5)):*
<details><summary><i>Previously, enitrat (Mathieu) wrote…</i></summary><blockquote>

"a variety of smart pointers"
"provide various functionalities"

this doesn't give any meaningful information
</blockquote></details>

yeah, removed that
___
*[`src/ch11-03-smart-pointers.md` line 11 at r2](https://reviewable.io/reviews/cairo-book/cairo-book/750#-NwTrMjJ6t26t9pOaU6q:-NwTy8TG4bteneTQgq7N:bkyj9q9) ([raw file](https://github.com/cairo-book/cairo-book/blob/fe1172b9b6e9a9570375296ee7d12add87cef05b/src/ch11-03-smart-pointers.md#L11)):*
<details><summary><i>Previously, enitrat (Mathieu) wrote…</i></summary><blockquote>

Very misleading imo.
</blockquote></details>

yes now i read this again its not good, its not its only purpose anyway and its not like you will use boxes to store any data (like a bool) in a safely manner. 
___
*[`src/ch11-03-smart-pointers.md` line 15 at r2](https://reviewable.io/reviews/cairo-book/cairo-book/750#-NwTrZHo22Ac8eBfd2gE:-NwU-caz-ek1HDC8XKuN:b-eva7sq) ([raw file](https://github.com/cairo-book/cairo-book/blob/fe1172b9b6e9a9570375296ee7d12add87cef05b/src/ch11-03-smart-pointers.md#L15)):*
> ```Markdown
> The `Box` type represents the main smart pointer type in Cairo. It allows us to store data in the `boxed_segment` segment of the Cairo VM memory. This segment is dedicated to the storage of all boxed values. Whenever you instantiate a new pointer variable of type `Box<T>`, you append the data of type `T` to the boxed segment.
> 
> Using boxes can improve efficiency as it allows to pass the variable, which is a pointer, from a function to another without the need to copy the whole value. Indeed, there is no access to the value contained in the box while there is no unbox. We still need to execute steps to copy the value of the pointer itself, but not the value the box points to. If the variable stored in the box is very complex, improvement can be massive. Because there is a dedicated segment for boxes and because a box can only be used in one place at a time, using the `Box<T>` type when manipulating complex data types reduces the number of steps while ensuring soundness in pointers dereferences.
> ```

way better
___
*[`src/ch11-03-smart-pointers.md` line 15 at r2](https://reviewable.io/reviews/cairo-book/cairo-book/750#-NwTrld2Ct_3aAKqO7Bl:-NwUDxAv7JJzWxKG1Jf_:b-896fix) ([raw file](https://github.com/cairo-book/cairo-book/blob/fe1172b9b6e9a9570375296ee7d12add87cef05b/src/ch11-03-smart-pointers.md#L15)):*
<details><summary><i>Previously, enitrat (Mathieu) wrote…</i></summary><blockquote>

Instead of having to write `n' values into memory before you call a function, you'd only have to write a single value that corresponds to your data pointer.
</blockquote></details>

Done.
___
*[`src/ch11-03-smart-pointers.md` line 15 at r2](https://reviewable.io/reviews/cairo-book/cairo-book/750#-NwTs8OR2BORLvWLsTuP:-NwUEjo-EnhXMNMYOYyR:b-73enwk) ([raw file](https://github.com/cairo-book/cairo-book/blob/fe1172b9b6e9a9570375296ee7d12add87cef05b/src/ch11-03-smart-pointers.md#L15)):*
<details><summary><i>Previously, enitrat (Mathieu) wrote…</i></summary><blockquote>

Confusing: you're confusing the safety (of the Box type) and the efficiency properties (of pointers)
</blockquote></details>

yeah agree, this was like an ''all in one'' try , not clear at all as it is placed after efficiency explanations and before safety . to be removed anyway 
___
*[`src/ch11-03-smart-pointers.md` line 22 at r2](https://reviewable.io/reviews/cairo-book/cairo-book/750#-NwTsXSa7rURIttSm-aq:-NwUY34M2kgJqhAWyQDa:b-8qjatt) ([raw file](https://github.com/cairo-book/cairo-book/blob/fe1172b9b6e9a9570375296ee7d12add87cef05b/src/ch11-03-smart-pointers.md#L22)):*
<details><summary><i>Previously, enitrat (Mathieu) wrote…</i></summary><blockquote>

do they?
</blockquote></details>

no they dont


<!-- Sent from Reviewable.io -->

TAdev0 commented:
@enitrat  it starts to look good. I added your short example of box declarations with CASM which is nice, thanks again for compilation optimisation issues i had.  i reworked a bit the second example so that :

- first example with box declaration highlights the memory layout with boxes
- const example highlights efficiency when passing box to functions

TAdev0 commented:
ok i can do better. 
- will remove the const thing, to just pass structs
- so that i can use 2.5.3 to build the non optimized casm code, because complex const were not supported at the time

enitrat commented:
you don't necessarily need consts, just use a tuple or a regular struct

enitrat reviewed (changes requested):
> Reviewed 5 of 7 files at r3, 1 of 6 files at r4, 6 of 9 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cairo-book/cairo-book/750)* status: 11 of 15 files reviewed, 1 unresolved discussion (waiting on @TAdev0)

___
*[`src/ch11-03-smart-pointers.md` line 62 at r5](https://reviewable.io/reviews/cairo-book/cairo-book/750#-Nx88PhW-PDcJYZGsq31:-Nx88PhW-PDcJYZGsq32:b-3qx0np) ([raw file](https://github.com/cairo-book/cairo-book/blob/4c20984edafbcd168b61bb59f5b80201868eb365/src/ch11-03-smart-pointers.md#L62)):*
> ```Markdown
> When passing data to a function, the entire data is copied in the last available memory cells right before the function call. Calling `pass_data`, will copy all 3 fields of `RandomStruct` to memory, while `pass_pointer` only requires the copy of the `new_box` pointer which is of size 1.
> 
> <!-- TODO: add illustrations rather than CASM Code -->
> ```

After reading everything I don't think it's a good idea to show actual CASM code. Illustrations should be more than enough to understand what's happening, too many details is counterproductive. 

You can convey the exact same idea without looking at actual code and just having well-designed schemas


<!-- Sent from Reviewable.io -->

TAdev0 commented:
@enitrat  your illustration is very nice.
its very clear. Maybe we can merge and open an issue for the section `Enabling Recursive Types with Nullable Boxes` ?
